src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -240,19 +232,38 @@ HRESULT STDMETHODCALLTYPE CorProfiler::AssemblyLoadFinished(AssemblyID assembly_
       << '.'_W
       << ToWSTRING(assembly_metadata.version.build);
 
-  // Check that Major.Minor.Build match the profiler version
-  if (ws.str() == ToWSTRING(PROFILER_VERSION)) {
-    Info("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed v", ws.str(), " matched profiler version v", PROFILER_VERSION);
-    managed_profiler_loaded_app_domains.insert(assembly_info.app_domain_id);
+  if (debug_logging_enabled) {
+    Debug("AssemblyLoadFinished: AssemblyName=", assembly_info.name, " AssemblyVersion=", ws.str(), ".", assembly_metadata.version.revision);
+  }
 
-    if (runtime_information_.is_desktop() && corlib_module_loaded &&
-          assembly_info.app_domain_id == corlib_app_domain_id) {
-      Info("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed was loaded domain-neutral");
-      managed_profiler_loaded_domain_neutral = true;
+  if (assembly_info.name == "Datadog.Trace.ClrProfiler.Managed"_W) {
+    // Check that Major.Minor.Build match the profiler version
+    if (ws.str() == ToWSTRING(PROFILER_VERSION)) {
+      Info("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed v", ws.str(), " matched profiler version v", PROFILER_VERSION);
+      managed_profiler_loaded_app_domains.insert(assembly_info.app_domain_id);
+
+      if (runtime_information_.is_desktop() && corlib_module_loaded) {
+        // Set the managed_profiler_loaded_domain_neutral flag whenever the managed profiler is loaded shared
+        if (assembly_info.app_domain_id == corlib_app_domain_id) {
+          Info("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed was loaded domain-neutral");
+          managed_profiler_loaded_domain_neutral = true;
+        }
+        // Set the managed_profiler_unsafe_to_instrument_domain_neutral flag whenever the profiler is not loaded shared
+        else {
+          Info("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed was not loaded domain-neutral");
+          managed_profiler_unsafe_to_instrument_domain_neutral = true;
+
+          // If the profiler was previously loaded domain-neutral, emit a warning pointing out this problematic failure mode
+          if (managed_profiler_loaded_domain_neutral) {
+            Warn("AssemblyLoadFinished: Datadog.Trace.ClrProfiler.Managed was NOT loaded domain-neutral. The CLR determined that one of the dependencies of Datadog.Trace.ClrProfiler.Managed was unable to load in the AppDomain \"EE Shared Assembly Repository\". Enable DD_TRACE_DEBUG to see this debug-level info.");
+            Warn("WARNING: The profiler in this process (in IIS this means the Application Pool) will stop instrumenting Framework assemblies in the GAC such as System.Data, System.Net.Http, etc. To fix this issue, ensure that all BindingRedirects for System.Net.Http and dependencies of Datadog.Trace.ClrProfiler.Managed result in assemblies found in the GAC.");
```

Comment:
Great messages here Zach.

Reply:
Thanks, but I sadly ended up stripping most of this detailed text because it no longer applies, as the decision now is to just never instrument domain-neutral assemblies, instead of basing the decision off the managed profiler